File: src/algorithms/strings/transformation/longest-common-prefix/sources/LongestCommonPrefix.cpp
// Longest Common Prefix — vertical scanning column by column across all strings.
// Returns the longest prefix shared by every word in the input array.
// Time: O(n*m) where n = number of strings, m = min string length  Space: O(1)

#include <string>
#include <vector>

std::string longestCommonPrefix(const std::vector<std::string>& words) {
    if (words.empty()) return ""; // @step:initialize

    int prefixLength = 0; // @step:initialize

    const std::string& firstWord = words[0]; // @step:initialize

    for (int columnIndex = 0; columnIndex < static_cast<int>(firstWord.length()); columnIndex++) {
        char currentChar = firstWord[columnIndex]; // @step:read-char

        for (int wordIndex = 1; wordIndex < static_cast<int>(words.size()); wordIndex++) {
            const std::string& word = words[wordIndex]; // @step:read-char
            char wordChar = (columnIndex < static_cast<int>(word.length())) ? word[columnIndex] : '\0'; // @step:read-char

            if (wordChar != currentChar) {
                return firstWord.substr(0, prefixLength); // @step:complete
            }
        }

        prefixLength++; // @step:write-char
    }

    return firstWord.substr(0, prefixLength); // @step:complete
}

File: src/algorithms/strings/transformation/longest-common-prefix/sources/longest-common-prefix.ts
// Longest Common Prefix — vertical scanning column by column across all strings.
// Returns the longest prefix shared by every word in the input array.
// Time: O(n*m) where n = number of strings, m = min string length  Space: O(1)

function longestCommonPrefix(words: string[]): string {
  if (words.length === 0) return ""; // @step:initialize

  let prefixLength = 0; // @step:initialize

  const firstWord = words[0] ?? ""; // @step:initialize

  for (let columnIndex = 0; columnIndex < firstWord.length; columnIndex++) {
    const currentChar = firstWord[columnIndex]; // @step:read-char

    for (let wordIndex = 1; wordIndex < words.length; wordIndex++) {
      const word = words[wordIndex] ?? ""; // @step:read-char
      const wordChar = word[columnIndex]; // @step:read-char

      if (wordChar !== currentChar) {
        return firstWord.slice(0, prefixLength); // @step:complete
      }
    }

    prefixLength++; // @step:write-char
  }

  return firstWord.slice(0, prefixLength); // @step:complete
}
